compiler/nir/nir.nim

```diff
@@ -61,11 +65,49 @@ proc runCode*(c: PPassContext; n: PNode): PNode =
     result = n
   c.oldErrorCount = c.m.graph.config.errorCounter
 
-when false:
-  type
-    Module* = object
-      types: TypeGraph
-      data: seq[Tree]
-      init: seq[Tree]
-      procs: seq[Tree]
+type
+  NirPassContext* = ref object of TPassContext
+    m: ModuleCon
+    c: ProcCon
+
+proc openNirBackend*(g: ModuleGraph; module: PSym; idgen: IdGenerator): PPassContext =
+  let m = initModuleCon(g, g.config, idgen, module)
+  NirPassContext(m: m, c: initProcCon(m, nil, g.config), idgen: idgen)
+
+proc gen(c: NirPassContext; n: PNode) =
+  let n = transformExpr(c.m.graph, c.idgen, c.m.module, n)
+  let pc = genStmt(c.c, n)
+
+proc nirBackend*(c: PPassContext; n: PNode): PNode =
+  gen(NirPassContext(c), n)
+  result = n
+
+proc closeNirBackend*(c: PPassContext; finalNode: PNode) =
+  discard nirBackend(c, finalNode)
+
+  let c = NirPassContext(c)
+  let outp = getNimcacheDir(c.c.config).string / c.m.module.name.s.addFileExt("nir")
```

Comment:
Create dir for this please, currently fails to create file unless dir already exists.

Reply:
Haven't decided yet how to mangle the name, the old way with @ is too ugly.